cli/src/commands/fix.rs

```diff
@@ -0,0 +1,153 @@
+use jj_lib::commit::{Commit, CommitIteratorExt};
+use jj_lib::matchers::EverythingMatcher;
+use jj_lib::merged_tree::MergedTreeBuilder;
+use jj_lib::repo::Repo;
+use jj_lib::repo_path::RepoPathBuf;
+use jj_lib::revset::{RevsetExpression, RevsetIteratorExt};
+use jj_lib::store::Store;
+use pollster::FutureExt;
+use tracing::instrument;
+
+use crate::cli_util::{CommandHelper, RevisionArg};
+use crate::command_error::CommandError;
+use crate::ui::Ui;
+
+/// Format files
+#[derive(clap::Args, Clone, Debug)]
+#[command(verbatim_doc_comment)]
+pub(crate) struct FixArgs {
+    /// Fix these commits and all their descendants
+    #[arg(long, short)]
+    sources: Vec<RevisionArg>,
+}
+
+#[instrument(skip_all)]
+pub(crate) fn cmd_fix(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &FixArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let root_commits: Vec<Commit> = workspace_command
+        .parse_union_revsets(&args.sources)?
+        .evaluate_to_commits()?
+        .try_collect()?;
+    workspace_command.check_rewritable(root_commits.iter().ids())?;
+
+    let mut tx = workspace_command.start_transaction();
+
+    // Collect all FileIds we're going to format and which commits they appear in
+    let commits: Vec<_> = RevsetExpression::commits(root_commits.iter().ids().cloned().collect())
+        .descendants()
+        .evaluate_programmatic(tx.base_repo().as_ref())?
+        .iter()
+        .commits(tx.repo().store())
+        .try_collect()?;
+    let mut file_ids = HashSet::new();
+    let mut commit_paths: HashMap<CommitId, Vec<RepoPathBuf>> = HashMap::new();
+    for commit in commits.iter().rev() {
+        let mut paths = vec![];
+        // Paths modified in parent commits in the set should also be updated in this
+        // commit
+        for parent_id in commit.parent_ids() {
+            if let Some(parent_paths) = commit_paths.get(parent_id) {
+                paths.extend_from_slice(parent_paths);
+            }
+        }
+        let parent_tree = commit.parent_tree(tx.repo())?;
+        let tree = commit.tree()?;
+        let mut diff_stream = parent_tree.diff_stream(&tree, &EverythingMatcher);
+        async {
+            while let Some((repo_path, diff)) = diff_stream.next().await {
+                let (_before, after) = diff?;
+                for term in after.into_iter().flatten() {
+                    if let TreeValue::File { id, executable: _ } = term {
+                        file_ids.insert((repo_path.clone(), id));
+                        paths.push(repo_path.clone());
+                    }
+                }
+            }
+            Ok::<(), BackendError>(())
+        }
+        .block_on()?;
+        commit_paths.insert(commit.id().clone(), paths);
+    }
+
+    let formatted = format_files(tx.repo().store().as_ref(), &file_ids)?;
+
+    tx.mut_repo().transform_descendants(
```

Comment:
comment: One thing to consider is how to recover from failures. My initial expectation was that I would want the entire process to fail if processing any commit failed. But, so far in my experience running `git test fix`, if processing one or more commits fails —

- I've still actually always wanted to apply any fixes for commits whose ancestors did not fail.
- In some but not all cases, I've wanted to apply fixes for commits that passed, even if some ancestor failed.
- I don't think I've ever wanted to apply a fix if the command for it returned a non-zero exit status, even if it did successfully format some files on disk.

Reply:
The approach that has worked well for us in `hg fix` is to only use the output if the exit code is zero, and to print any stderr lines with a prefix that indicates which commit/file caused it. Part of this problem is just meeting the formatting tools where they are instead of trying to negotiate some new contract with them.